crypto/rand/rand_vms.c

```diff
@@ -497,6 +508,73 @@ int rand_pool_add_nonce_data(RAND_POOL *pool)
     return rand_pool_add(pool, (unsigned char *)&data, sizeof(data), 0);
 }
 
+/*
+ * SYS$GET_ENTROPY METHOD
+ * ======================
+ *
+ * This is a high entropy method based on a new systems service that is
+ * based on getentropy() from FreeBSD 12.  It's only used if available,
+ * and its availability is detected in run-time.
+ *
+ * We assume that this function gives 1 bit of entropy for each bit of
+ * output.
+ */
+#define PUBLIC_VECTORS "SYS$PUBLIC_VECTORS"
+#define GET_ENTROPY "SYS$GET_ENTROPY"
+
+static int get_entropy_address_flag = 0;
+static int (*get_entropy_address)(void *buffer, size_t buffer_size) = NULL;
+static int init_get_entropy_address(void)
+{
+    if (get_entropy_address_flag == 0) {
+        get_entropy_addr = dlsym(dlopen(PUBLIC_VECTORS, 0), OLD_ROUTINE);
+    }
+    get_entropy_address_flag = 1;
+    return get_entropy_addr != NULL;
+}
+
+size_t get_entropy_method(RAND_POOL *pool)
+{
+    /*
+     * Documentation says that SYS$GET_ENTROPY will give a maximum of
+     * 256 bytes of data.
+     */
+    unsigned char buffer[256];
+    size_t bytes_needed = rand_pool_bytes_needed(pool, 1);
```

Comment:
Is a range check against 256 appropriate?
I realise that 2048 bits seems unlikely....

Reply:
I use `bytes_remaining` in the call to the system service, and if that's too big, it will return a status that says so (i.e. Something other than `SS$_NORMAL` or `SS$_RETRY`.